api-client/src/main/java/com/takipi/api/core/url/UrlClient.java

```diff
@@ -6,34 +6,92 @@
 import java.io.Reader;
 import java.net.HttpURLConnection;
 import java.net.URL;
+import java.util.List;
 
 import org.apache.commons.io.IOUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import com.google.common.base.Strings;
+import com.google.common.collect.Lists;
 import com.takipi.api.core.consts.ApiConstants;
 import com.takipi.common.util.Pair;
 
 public abstract class UrlClient {
+	
+	public enum Operation {GET, DELETE, POST, PUT } 
+	
+	public interface UrlClientObserver {
+		public void observe(Operation operation, String url, String response, long time);
+	}
+	
 	protected static final Logger logger = LoggerFactory.getLogger(UrlClient.class);
 
 	static final Response<String> BAD_RESPONSE = Response.of(HttpURLConnection.HTTP_INTERNAL_ERROR, null);
 
 	private final String hostname;
 	private final int connectTimeout;
 	private final int readTimeout;
-
+	private final Object observerLock;
+	private volatile List<UrlClientObserver> observers;
+	
 	protected UrlClient(String hostname, int connectTimeout, int readTimeout) {
 		this.hostname = hostname;
 		this.connectTimeout = connectTimeout;
 		this.readTimeout = readTimeout;
+		this.observerLock = new Object();
+	}
+	
+	private void iterateObservers(Operation operation, String url, String response, long time) {
+		
+		List<UrlClientObserver> observers = this.observers;
+		
+		if (observers == null) {
+			return;
+		}
+		
+		for (UrlClientObserver observer : observers) {
+			observer.observe(operation, url, response, time);
+		}
+	}
+
+	public void addObserver(UrlClientObserver observer) {
+		
+		if (observer == null) {
+			throw new IllegalArgumentException("observer");
+		}
+		
+		synchronized (observerLock)
+		{
+			if (observers != null) {
+				observers = Lists.newArrayList(observers);
+			} else {
+				observers = Lists.newArrayList();
+			}
+			
+			observers.add(observer);
```

Comment:
Because `iterateObservers` is (rightfully) not protected by the synchronized block, you risk modifying the observers list between the new list creation and the addition performed in this line.

Can be easily resolved by the creation being done to a temp var, then add, and then assign into `this.observers`